### PR TITLE
Support graceful shutdowns

### PR DESCRIPTION
**Pull request by sidwebworks**
## Bug

- [ ] Related issues linked using `fixes #number`
- [ ] Integration tests added
- [ ] Errors have helpful link attached, see `contributing.md`

## Feature

- [ ] Implements an existing feature request or RFC. Make sure the feature request has been accepted for implementation before opening a PR.
- [x] Related issues linked using [19693](https://github.com/vercel/next.js/discussions/19693)
- [ ] Integration tests added
- [ ] Documentation added
- [ ] Telemetry added. In case of a feature if it's used or not.
- [ ] Errors have helpful link attached, see `contributing.md`

## Documentation / Examples

- [ x] Make sure the linting passes by running `yarn lint`

Closes #29959

### CODE REVIEW

File: packages/next/bin/next.ts [hunk 1]
```diff
@@ -129,7 +129,16 @@ if (process.versions.pnp === '3') {
...
 
 // Make sure commands gracefully respect termination signals (e.g. from Docker)
-process.on('SIGTERM', () => process.exit(0))
+// Allow the graceful termination to be manually configurable
+if (!process.env.NEXT_MANUAL_SIGTERM) {
+  process.on('SIGTERM', () =>
+    setTimeout(
+      () => process.exit(0),
+      parseInt(process.env.NEXT_GRACEFUL_SHUTDOWN_TIMEOUT_MS || '0', 10)
+    )
```
Comment (SukkaW) [inline packages/next/bin/next.ts:138]: Why `setTimeout` though? What about `process.nextTick`?

File: packages/next/bin/next.ts [hunk 2]
```diff
@@ -129,7 +129,16 @@ if (process.versions.pnp === '3') {
 }
 
 // Make sure commands gracefully respect termination signals (e.g. from Docker)
-process.on('SIGTERM', () => process.exit(0))
+// Allow the graceful termination to be manually configurable
+if (!process.env.NEXT_MANUAL_SIGTERM) {
+  process.on('SIGTERM', () =>
+    setTimeout(
+      () => process.exit(0),
+      parseInt(process.env.NEXT_GRACEFUL_SHUTDOWN_TIMEOUT_MS || '0', 10)
```
Comment (ijjk) [inline packages/next/bin/next.ts:137]: This timeout shouldn't be added, if `NEXT_MANUAL_SIGTERM` is set, the user should add their own `SIGTERM` handling removing the need for an arbitrary timeout like mentioned in the requested changes here https://github.com/vercel/next.js/pull/29959#discussion_r803265324

Comment (sidwebworks) [reply packages/next/bin/next.ts:137]: Makes sense I will make this change, I left it so even if the user doesn't add a manual graceful shutdown handler in the `_document ` but still wants to add an arbitiary delay.

 Now that I think about it sounds kinda like a stupid inconsistency. 😆

File: packages/next/bin/next.ts (hunk 1 above)
Comment (sidwebworks) [reply packages/next/bin/next.ts:138]: Ah nevermind we would we better off removing the timeout all together, as the user can handle the graceful shutdown however they want it to be.  

Sorry I know it looks like I wanted it to exit out immediately as `0` is passed to the timeout and `process.nextTick` would've prioritized this before than the timeout, but it was more like funny customization.

File: packages/next/bin/next.ts (hunk 2 above)
Comment (ijjk) [reply packages/next/bin/next.ts]: I see, let's also include the SIGINT in the env variable check, might make sense to call it something like NEXT_MANUAL_SIG_HANDLE

And we can also mention this env variable in the deployment doc here https://github.com/vercel/next.js/blob/canary/docs/deployment.md#nodejs-server

Comment (sidwebworks) [reply packages/next/bin/next.ts]:
```js
if (!process.env.NEXT_MANUAL_SIG_HANDLE) {
  process.on('SIGTERM', () => process.exit(0))
  process.on('SIGINT', () => process.exit(0))
}
```
Like this?

Comment (ijjk) [reply packages/next/bin/next.ts]: Yeah exactly

Comment (sidwebworks) [reply packages/next/bin/next.ts]: Cool, will update the docs now

Comment (sidwebworks) [reply packages/next/bin/next.ts]: Done

File: docs/deployment.md [hunk 3]
```diff
@@ -96,7 +96,28 @@ When you deploy your Next.js application, you want to see the latest version wit
 
 Next.js will automatically load the latest version of your application in the background when routing. For client-side navigations, `next/link` will temporarily function as a normal `<a>` tag.
 
-**Note:** If a new page (with an old version) has already been prefetched by `next/link`, Next.js will use the old version. Navigating to a page that has _not_ been prefetched (and is not cached at the CDN level) will load the latest version.
```
Comment (balazsorban44) [inline docs/deployment.md:99]: Was this section removed by accident?

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | sidwebworks/next.js canary | Change |
| - | - | - | - |
| buildDuration | 17.8s | 19.3s | ⚠️ +1.5s |
| buildDurationCached | 7s | 7s | -20ms |
| nodeModulesSize | 475 MB | 475 MB | ⚠️ +364 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | sidwebworks/next.js canary | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 4.852 | 5.506 | ⚠️ +0.65 |
| / avg req/sec | 515.23 | 454.09 | ⚠️ -61.14 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 2.104 | 2.217 | ⚠️ +0.11 |
| /error-in-render avg req/sec | 1188.03 | 1127.49 | ⚠️ -60.54 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | sidwebworks/next.js canary | Change |
| - | - | - | - |
| 925.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42 kB | 42 kB | ✓ |
| main-HASH.js gzip | 28.6 kB | 28.6 kB | ✓ |
| webpack-HASH.js gzip | 1.44 kB | 1.44 kB | ✓ |
| Overall change | 72.3 kB | 72.3 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | sidwebworks/next.js canary | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | sidwebworks/next.js canary | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.36 kB | 1.36 kB | ✓ |
| _error-HASH.js gzip | 193 B | 193 B | ✓ |
| amp-HASH.js gzip | 308 B | 308 B | ✓ |
| css-HASH.js gzip | 327 B | 327 B | ✓ |
| dynamic-HASH.js gzip | 2.7 kB | 2.7 kB | ✓ |
| head-HASH.js gzip | 359 B | 359 B | ✓ |
| hooks-HASH.js gzip | 920 B | 920 B | ✓ |
| image-HASH.js gzip | 5.71 kB | 5.71 kB | ✓ |
| index-HASH.js gzip | 263 B | 263 B | ✓ |
| link-HASH.js gzip | 2.65 kB | 2.65 kB | ✓ |
| routerDirect..HASH.js gzip | 320 B | 320 B | ✓ |
| script-HASH.js gzip | 391 B | 391 B | ✓ |
| withRouter-HASH.js gzip | 318 B | 318 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 15.9 kB | 15.9 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | sidwebworks/next.js canary | Change |
| - | - | - | - |
| _buildManifest.js gzip | 458 B | 458 B | ✓ |
| Overall change | 458 B | 458 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | sidwebworks/next.js canary | Change |
| - | - | - | - |
| index.html gzip | 532 B | 532 B | ✓ |
| link.html gzip | 544 B | 544 B | ✓ |
| withRouter.html gzip | 527 B | 527 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Default Build with SWC</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | sidwebworks/next.js canary | Change |
| - | - | - | - |
| buildDuration | 20.1s | 20.1s | -77ms |
| buildDurationCached | 6.9s | 7.1s | ⚠️ +186ms |
| nodeModulesSize | 475 MB | 475 MB | ⚠️ +364 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | sidwebworks/next.js canary | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 4.929 | 5.375 | ⚠️ +0.45 |
| / avg req/sec | 507.17 | 465.07 | ⚠️ -42.1 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.937 | 1.966 | ⚠️ +0.03 |
| /error-in-render avg req/sec | 1290.53 | 1271.48 | ⚠️ -19.05 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | sidwebworks/next.js canary | Change |
| - | - | - | - |
| 925.HASH.js gzip | 178 B | 178 B | ✓ |
| framework-HASH.js gzip | 42.2 kB | 42.2 kB | ✓ |
| main-HASH.js gzip | 29.1 kB | 29.1 kB | ✓ |
| webpack-HASH.js gzip | 1.45 kB | 1.45 kB | ✓ |
| Overall change | 73 kB | 73 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | sidwebworks/next.js canary | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | sidwebworks/next.js canary | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.35 kB | 1.35 kB | ✓ |
| _error-HASH.js gzip | 179 B | 179 B | ✓ |
| amp-HASH.js gzip | 312 B | 312 B | ✓ |
| css-HASH.js gzip | 324 B | 324 B | ✓ |
| dynamic-HASH.js gzip | 2.72 kB | 2.72 kB | ✓ |
| head-HASH.js gzip | 357 B | 357 B | ✓ |
| hooks-HASH.js gzip | 921 B | 921 B | ✓ |
| image-HASH.js gzip | 5.8 kB | 5.8 kB | ✓ |
| index-HASH.js gzip | 261 B | 261 B | ✓ |
| link-HASH.js gzip | 2.77 kB | 2.77 kB | ✓ |
| routerDirect..HASH.js gzip | 322 B | 322 B | ✓ |
| script-HASH.js gzip | 392 B | 392 B | ✓ |
| withRouter-HASH.js gzip | 317 B | 317 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 16.1 kB | 16.1 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | sidwebworks/next.js canary | Change |
| - | - | - | - |
| _buildManifest.js gzip | 458 B | 458 B | ✓ |
| Overall change | 458 B | 458 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | sidwebworks/next.js canary | Change |
| - | - | - | - |
| index.html gzip | 528 B | 528 B | ✓ |
| link.html gzip | 543 B | 543 B | ✓ |
| withRouter.html gzip | 525 B | 525 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
Commit: 8f0540d6f81c5b59d3f79bd8c68cf61d5b72cbc2

File: docs/deployment.md (hunk 3 above)
Comment (sidwebworks) [reply docs/deployment.md:99]: Ah yes I think so, will undo this.

**Review by ijjk — APPROVED** [review]
Thanks for the PR!

**Pull request merged**